cpp/src/io/utilities/trie.cuh

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2018-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @brief Serialized trie implementation for C++/CUDA
+ * @file trie.cuh
+ */
+
+#pragma once
+
+#include <cudf/utilities/span.hpp>
+
+namespace cudf {
+namespace detail {
+static constexpr char trie_terminating_character = '\n';
+
+struct serial_trie_node {
+  int16_t children_offset{-1};
+  char character{trie_terminating_character};
+  bool is_leaf{false};
+  // serial_trie_node() = default;  // FIXME This is necessary for a Thrust bug on CentOS7 + CUDA10
+  explicit serial_trie_node(char c, bool leaf = false) noexcept : character(c), is_leaf(leaf) {}
+};
+
+using trie          = rmm::device_uvector<serial_trie_node>;
+using optional_trie = std::optional<trie>;
+using trie_view     = device_span<serial_trie_node const>;
+
+inline trie_view make_trie_view(optional_trie const& t)
+{
+  if (!t) return {};
+  return trie_view{t->data(), t->size()};
+}
+
+/**
+ * @brief Creates a serialized trie for cache-friendly string search.
+ *
+ * The resulting trie is a compact array - children array size is equal to the
+ * actual number of children nodes, not the size of the alphabet.
+ *
+ * @param keys Array of strings to insert into the trie
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ *
+ * @return A host vector of nodes representing the serialized trie
+ */
+trie create_serialized_trie(const std::vector<std::string>& keys, rmm::cuda_stream_view stream);
+
+/*
+ * @brief Searches for a string in a serialized trie.
+ *
+ * Can be executed on host or device, as long as the data is available
+ *
+ * @param trie Pointer to the array of nodes that make up the trie
+ * @param key Pointer to the start of the string to find
+ * @param key_len Length of the string to find
+ *
+ * @return Boolean value; true if string is found, false otherwise
+ */
+__host__ __device__ inline bool serialized_trie_contains(device_span<serial_trie_node const> trie,
```

Comment:
I don't think a `device_span` can be passed to `__device__` code. Does this function really need to be `__host__ __device__`?

Reply:
It can be passed to kernels, we use this a lot in cuIO.
This function is used in CSV and JSON kernels to filter special values, so it needs to be __device__.

Reply:
OK. I think device_span works in device code because or relaxed `constexpr`.  So we should watch out for the problems discussed in #7795